api/v1alpha1/stage_types.go

```diff
@@ -155,15 +155,15 @@ type ImageSubscription struct {
 	// value in this field MUST NOT include an image tag. This field is required.
 	//
 	//+kubebuilder:validation:MinLength=1
-	//+kubebuilder:validation:Pattern=`^(([\w\d\.-]+)(:[\d]+)?/)?[a-z0-9-]+(/[a-z0-9-]+)*$`
+	//+kubebuilder:validation:Pattern=`^(([\w\d\.-]+)(:[\d]+)?/)?[a-z0-9-_\.]+(/[a-z0-9-_\.]+)*$`
 	RepoURL string `json:"repoURL"`
 	// GitRepoURL optionally specifies the URL of a Git repository that contains
 	// the source code for the image repository referenced by the RepoURL field.
 	// When this is specified, Kargo MAY be able to infer and link to the exact
 	// revision of that source code that was used to build the image.
 	//
 	//+kubebuilder:validation:Optional
-	//+kubebuilder:validation:Pattern=`^(((https?://)|([\w-]+@))([\w\d\.]+)(:[\d]+)?/(.*))?$`
+	//+kubebuilder:validation:Pattern=`^((https?://)|([\w.-@]+@))([\w.-]+)(:[\d]+)?(/(.*))?$`
```

Comment:
I think your new expression is a clear improvement over the old one, but I think there are some other issues that should be corrected while we're already touching this.

tbh, I probably copied and pasted this regex from somewhere else in the very early days of Kargo.

The system doesn't actually support http or ssh yet, so while parts of this need to be relaxed, the far left side needs to be tightened up until such time that we might add http or ssh support.

Here's what I propose: ^https://(\w+([\.-]\w+)*@)?\w+([\.-]\w+)*(:[\d]+)?(/.*)?$

Broken down:

| Expression Part | Explanation |
|-----------------|-------------|
| ^ | |
| https:// | HTTPS only |
| (\w+([\.-]\w+)*@)? | Optional token@ before domain name; must start and end with a word character |
| \w+([\.-]\w+)* | Domain name; must start and end with a word character |
| (:[\d]+)? | Optional port number |
| (/.*)? | Anything goes after the first / |
| $ | |

![Screenshot 2023-10-04 at 11 27 02 AM](https://github.com/akuity/kargo/assets/1821014/fd8533ec-81be-4d1e-9591-55fd49df707c)

What do you think about this? It does retain your improvements.

Also applicable on line 266.

Reply:
I tried to get fancy squashing inside my branch so your comments got disassociated - but I accepted all your suggestions.